### PR TITLE
Jaa/is nilpotent

### PR DESCRIPTION
**Pull request by JohnAAbbott**
Usefully faster implementation of `is_nilpotent(::ResElem)`, but it is longer...

### CODE REVIEW

File: src/flint/fmpq.jl [hunk 1]
```diff
@@ -115,6 +115,8 @@ is_zero(a::QQFieldElemOrPtr) = is_zero(_num_ptr(a))
 
 is_unit(a::QQFieldElem) = !iszero(a)
 
+is_nilpotent(a::QQFieldElem) = iszero(a)
```
Comment (fingolfin) [inline src/flint/fmpq.jl:118]: I am not sure we really need or want these methods here -- I would expect this to be handled by a generic method in AA, something like this:
```
function is_nilpotent(a::T) where {T <: RingElem}
  is_domain_type(parent_type(T)) && return izero(a)
  throw(NotImplementedError(:is_nilpotent, a))
end
```

In general I think it would be better to first have the generic code in AA ready. BTW you mentioned again during the call today the problem of dealing with infinite series -- so I'll also mention again that this is not necessary for now -- the above method takes care of it well enough for now.

So a first PR just adding e.g. that method and perhaps a few tests would already be a good start. Adding `is_unit` methods can then come in one more further PRs. And then once AA is ready and released we can add missing or optimized bits in Nemo.

File: src/flint/fmpz_mod.jl [hunk 2]
```diff
@@ -20,6 +20,8 @@ base_ring(a::ZZModRing) = ZZ
 
 parent(a::ZZModRingElem) = a.parent
 
+is_domain_type(::Type{ZZModRingElem}) = false
```
Comment (fingolfin) [inline src/flint/fmpz_mod.jl:23]: This is not necessary as we have this in AA:
```
# Type can only represent elements of domains
# false unless explicitly specified
is_domain_type(R::Type{T}) where T <: RingElem = false
```

File: src/HeckeMoreStuff.jl [hunk 3]
```diff
@@ -391,11 +391,19 @@ end
...
+  # Code below is faster than
+  #    m=modulus(res);
+  #    return powermod(lift(res),nbits(m),m)==0
+  m = modulus(res)
+  r = lift(res)
+  while true
+    g = gcd(r, m)
+    (g == m) && return true
+    (g == 1) && return false
+    m /= g
```
Comment (thofma) [inline src/HeckeMoreStuff.jl:404]:
```suggestion
    m = divexact(m, g)
```
Since e.g. `/` is doing the wrong thing if `T === Int`:
```
julia> Int(2)/Int(3)
0.6666666666666666
```

Comment (JohnAAbbott) [reply src/HeckeMoreStuff.jl:404]: Bother!  I was hoping to do "in place" arithmetic with `m /= g` but I suspect that `m = divexact(m,g)` is not in place.  Too bad, I guess.  I'll check what the time penalty is...

Comment (fingolfin) [reply src/HeckeMoreStuff.jl:404]: `m /= g` is just syntactic sugar for `m = m / g` and does not work in place anyway.

We have `divexact!` for a few types to support in-place, but AFAIK this is still mostly in Nemo and and not yet generically implemented in AA. That could be subjective of a future PR. For now just use `divexact` ?

Comment (JohnAAbbott) [reply src/HeckeMoreStuff.jl:404]: **Question:**  in `AbstractAlgebra/src/Residue.jl` near line 73 there is the function definition
```
lift(a::ResElem{Int}) = BigInt(data(a))
```
Why is the result cast to a `BigInt`?  I had expected the result to be of type `Int` or maybe `UInt`; and if the result is to be cast to a "big integer" then wouldn't `ZZRingElem` be more natural (but then it seems that `ZZRingElem` is not visible inside `AbstractAlgebra`... sigh!)

Comment (JohnAAbbott) [reply src/HeckeMoreStuff.jl:404]: `divexact` was faster -- good!  I shall replace the call to `lift` by a call to `data` so that the code actually does what the reader might reasonably expect.  I shall also add an overflow test.

Comment (JohnAAbbott) [reply src/HeckeMoreStuff.jl]: I've pushed all the changes... waiting for GitHub to find my mistakes...

File: src/HeckeMoreStuff.jl [hunk 4]
```diff
@@ -391,11 +391,19 @@ end
...
+  #    m=modulus(res);
+  #    return powermod(lift(res),nbits(m),m)==0
+  m = modulus(res)
+  r = data(res) # the least non-negative class representative as a value of type "unsigned" T; !!note that "lift" casts the value to BigInt!!
+  while true
+    g = gcd(r, m)
+    (g == m) && return true
+    (g == 1) && return false
+    m = divexact(m, g)  #  equiv to:  m /= g
+    mod!(g, m);  r = g^2  # g^2 cannot overflow thanks to mod!
```
Comment (fingolfin) [inline src/HeckeMoreStuff.jl:405]:
```suggestion
    mod!(g, m)  r = g^2  # g^2 cannot overflow thanks to mod!
```

File: src/flint/nmod.jl [hunk 5]
```diff
@@ -20,6 +20,7 @@ base_ring(a::zzModRing) = ZZ
 
 parent(a::zzModRingElem) = a.parent
 
+
```
Comment (fingolfin) [inline src/flint/nmod.jl:23]:
```suggestion
```

File: src/flint/nmod_poly.jl [hunk 6]
```diff
@@ -20,6 +20,7 @@ elem_type(::Type{zzModPolyRing}) = zzModPolyRingElem
 
 dense_poly_type(::Type{zzModRingElem}) = zzModPolyRingElem
 
+
```
Comment (fingolfin) [inline src/flint/nmod_poly.jl:23]:
```suggestion
```

**Comment by JohnAAbbott** [comment]
The new code (or most of it) has been moved into `AbstractAlgebra`.

**Comment by JohnAAbbott** [comment]
**Bother!**  This PR will work only with the revised AA (see PR 1933 there)

File: test/flint/fmpz_mod-test.jl [hunk 7]
```diff
@@ -128,6 +131,21 @@ end
   @test_throws Exception R6(R22(1))
   @test_throws Exception R2(R3(1))
   @test_throws Exception R3(R2(1))
+
+  ZZmod720,_ = residue_ring(ZZ,720)
```
Comment (fingolfin) [inline test/flint/fmpz_mod-test.jl:135]: This is a different ring, so the test for that one should be in `nmod-test.jl`.

We also generally place a space after commas in our code

File: test/flint/fmpz_mod-test.jl [hunk 8]
```diff
@@ -128,6 +131,21 @@ end
...
+  ZZmodZZ720,_ = residue_ring(ZZ,ZZ(720))
+  @test is_nilpotent(ZZmod720(30))
+  @test is_nilpotent(ZZmodZZ720(30))
+  nilp720 = filter((r -> is_nilpotent(ZZmod720(r))), 0:719);
+  @test length(nilp720) == 24  # 720/(2*3*5)
```
Comment (fingolfin) [inline test/flint/fmpz_mod-test.jl:140]: Better expressed like so:
```suggestion
  @test count(is_nilpotent, ZZmod720) == 24  # 720/(2*3*5)
```

File: test/flint/fmpz_mod-test.jl [hunk 9]
```diff
@@ -128,6 +131,21 @@ end
...
+  nilp720 = filter((r -> is_nilpotent(ZZmodZZ720(r))), 0:719);
+  @test length(nilp720) == 24  # 720/(2*3*5)
+
+  # 64-bit overflow test:  **ASSUMES** Int is 64 bits
+  rr = 4*3^20
+  mm = 3*rr
+  R_overflow,_ = residue_ring(ZZ, mm)
+  @test is_nilpotent(R_overflow(rr))
```
Comment (fingolfin) [inline test/flint/fmpz_mod-test.jl:148]: What exactly is this testing? Clearly it is related to `Int` overflowing, but which aspect of that / where is it coming into play here?

File: src/flint/fmpq.jl (hunk 1 above)
Comment (fingolfin) [inline src/flint/fmpq.jl:118]:
```suggestion
```

File: src/flint/fmpz.jl [hunk 10]
```diff
@@ -254,6 +254,8 @@ size(a::ZZRingElem) = _fmpz_is_small(a) ? 1 : abs(_fmpz_size(a))
 
 is_unit(a::ZZRingElemOrPtr) = data(a) == 1 || data(a) == -1
 
+is_nilpotent(a::ZZRingElemOrPtr) = (data(a) == 0)
+
```
Comment (fingolfin) [inline src/flint/fmpz.jl:258]:
```suggestion
```

File: src/flint/nmod_mpoly.jl [hunk 11]
```diff
@@ -100,9 +100,9 @@ for (etype, rtype, ftype, ctype, utype) in (
       return length(a) == 1
     end
 
-    function is_unit(a::($etype))
-      return length(a) == 1 && total_degree(a) == 0 && is_unit(coeff(a, 1))
-    end
+    # function is_unit(a::($etype))
+    #   return length(a) == 1 && total_degree(a) == 0 && is_unit(coeff(a, 1))
+    # end
```
Comment (fingolfin) [inline src/flint/nmod_mpoly.jl:106]:
```suggestion
```

File: test/flint/fmpz_mod-test.jl (hunk 9 above)
Comment (JohnAAbbott) [reply test/flint/fmpz_mod-test.jl:148]: The current implementation of `is_nilpotent` works correctly for this input, but an earlier version was incorrect (even though the code looked right).  I prefer to keep this test example in case someone in the future modifies the code, and mishandles integer overflow.  Maybe this should be in `nmod-test` too?

**Comment by JohnAAbbott** [comment]
I am wondering whether all tests related to `is_nilpotent` should be moved into `AbstractAlgebra` since the implementation now resides there.  The distinction between what `AbstractAlgebra` handles and what `Nemo` handles is not entirely clear to me.
Maybe some tests need to be duplicated?  If I load the package `AbstractAlgebra`, there is already a function to create `residue_ring(ZZ,720)`; but if I load `Nemo` (or all of `Oscar`) as well then `residue_ring(ZZ,720)` calls a different function, presumably using a different interna; representation.  So probably the tests should be made both in the `AbstractAlgebra` tests and in the `Nemo` tests.

File: test/flint/fmpz_mod-test.jl (hunk 8 above)
Comment (JohnAAbbott) [reply test/flint/fmpz_mod-test.jl:140]: Your suggestion to use `count` works in an `Oscar` context but not in an `AbstractAlgebra` context because the following error arises:
```
julia> count(is_nilpotent, ZZmod720)
ERROR: MethodError: no method matching iterate(::EuclideanRingResidueRing{BigInt})
```
We could leave the test as is (_i.e._ using `filter`), or we could implement `iterate` for `EuclideanRingResidueRing{BigInt}`.  Which is the preferred/better option?

Comment (fingolfin) [reply test/flint/fmpz_mod-test.jl:140]: It doesn't work in AA, true, but it works fine in Nemo for me, and this PR is targetting Nemo, isn't it?

File: test/flint/fmpz_mod-test.jl (hunk 9 above)
Comment (fingolfin) [reply test/flint/fmpz_mod-test.jl:148]: The problem I have with this test is not the test itself (I am fine with having such a test) but rather the fact that it is not sufficiently documented, IMO. It suggests that something somewhere was presumably wrong at some point due to some overflow.

It seems 3^20 is reasonably close to `sqrt(2^63)=2^{31.5}` so hence `rr` and `mm` have squares that overflow an `Int64` and perhaps this is why it makes for a good test?!

**Comment by fingolfin** [comment]
To summarize what I explained earlier in my office: ideally as much of this testing as possible is done in generic "ring conformance tests" which then can be applied to all ring types we define in AA, Nemo, Hecke, Oscar, or elsewhere. In some cases one may still wish to have additional tests to e.g. cover certain corner cases or otherwise specific quirks of an implementation, these then of course should go into the respective package.

From that perspective the tests in here are not very special, and could just be dropped here.

**Comment by JohnAAbbott** [comment]
Hoping this is now fully complete.  The main implementations of `is_unit` and `is_nilpotent` for quotient rings of `ZZ` are now in `AbstractAlgebra`, but we do some tests here because there are new specialized implementations of quotient rings of `ZZ`.

File: src/flint/fmpz_mod.jl (hunk 2 above)
Comment (JohnAAbbott) [reply src/flint/fmpz_mod.jl]: Done
